File: iosApp/iosApp/Pages/Map/StopLayerGenerator.swift
//
//  StopLayerGenerator.swift
//  iosApp
//
//  Created by Simon, Emma on 3/25/24.
//  Copyright © 2024 MBTA. All rights reserved.
//

import shared
import SwiftUI
@_spi(Experimental) import MapboxMaps

class StopLayerGenerator {
    let stopLayers: [SymbolLayer] = createStopLayers()

    static let maxTransferLayers = 3
    static let stopZoomThreshold = 8.0

    static let stopLayerId = "stop-layer"
    static let stopTouchTargetLayerId = "\(stopLayerId)-touch-target"
    static let stopLayerSelectedPinId = "\(stopLayerId)-selected-pin"

    static let busLayerId = "\(stopLayerId)-bus"
    static let busAlertLayerId = "\(stopLayerId)-bus-alert"

    static func getAlertLayerId(_ index: Int) -> String {
        "\(stopLayerId)-alert-\(index.description)"
    }

    static func getTransferLayerId(_ index: Int) -> String {
        "\(stopLayerId)-transfer-\(index.description)"
    }

    static func createStopLayers() -> [SymbolLayer] {
        let sourceId = StopFeaturesBuilder.shared.stopSourceId
        let stopLayer = createStopLayer(id: Self.stopLayerId)
        let busLayer = createStopLayer(id: Self.busLayerId, forBus: true)

        var stopTouchTargetLayer = SymbolLayer(id: Self.stopTouchTargetLayerId, source: sourceId)
        stopTouchTargetLayer.iconImage = .expression(Exp(.image) { StopIcons.stopDummyIcon })
        stopTouchTargetLayer.iconPadding = .constant(22.0)
        includeSharedProps(on: &stopTouchTargetLayer)

        var stopSelectedPinLayer = SymbolLayer(id: Self.stopLayerSelectedPinId, source: sourceId)
        stopSelectedPinLayer.iconImage = .expression(Exp(.switchCase) {
            MapExp.shared.selectedExp.toMapbox()
            Exp(.image) { StopIcons.stopPinIcon }
            Exp(.image) { "" }
        })
        stopSelectedPinLayer.iconOffset = offsetPinValue()
        includeSharedProps(on: &stopSelectedPinLayer)

        let transferLayers = (0 ..< Self.maxTransferLayers).map { index in
            var transferLayer = SymbolLayer(id: Self.getTransferLayerId(index), source: sourceId)
            transferLayer.iconImage = StopIcons.getTransferLayerIcon(index)
            transferLayer.iconOffset = offsetTransferValue(index: index)
            includeSharedProps(on: &transferLayer)

            return transferLayer
        }

        let alertLayers = (0 ..< Self.maxTransferLayers).map { index in
            createAlertLayer(id: Self.getAlertLayerId(index), index: index)
        }

        let busAlertLayer = createAlertLayer(id: busAlertLayerId, forBus: true)

        return [stopTouchTargetLayer, busLayer, busAlertLayer, stopLayer] + transferLayers +
            alertLayers + [stopSelectedPinLayer]
    }

    static func createAlertLayer(id: String, index: Int = 0, forBus: Bool = false) -> SymbolLayer {
        var alertLayer = SymbolLayer(id: id, source: StopFeaturesBuilder.shared.stopSourceId)
        alertLayer.iconImage = AlertIcons.getAlertLayerIcon(index, forBus: forBus)
        alertLayer.iconOffset = offsetAlertValue(index: index)
        alertLayer.iconAllowOverlap = .constant(true)
        includeSharedProps(on: &alertLayer)

        return alertLayer
    }

    static func createStopLayer(id: String, forBus: Bool = false) -> SymbolLayer {
        var stopLayer = SymbolLayer(id: id, source: StopFeaturesBuilder.shared.stopSourceId)
        stopLayer.iconImage = StopIcons.getStopLayerIcon(forBus: forBus)
        stopLayer.textField = .expression(MapExp.shared.stopLabelTextExp(forBus: forBus))

        stopLayer.textColor = .constant(.init(.text))
        stopLayer.textFont = .constant(["Inter Regular"])
        stopLayer.textHaloColor = .constant(.init(.fill3))
        stopLayer.textHaloWidth = .constant(2.0)
        stopLayer.textSize = .constant(13)
        stopLayer.textVariableAnchor = .constant([.right, .bottom, .top, .left])
        stopLayer.textJustify = .constant(.auto)
        stopLayer.textAllowOverlap = .constant(true)
        stopLayer.textOptional = .constant(true)
        stopLayer.textOffset = .expression(MapExp.shared.labelOffsetExp.toMapbox())

        includeSharedProps(on: &stopLayer)
        return stopLayer
    }

    static func includeSharedProps(on layer: inout SymbolLayer) {
        layer.iconSize = .expression(MapExp.shared.selectedSizeExp.toMapbox())

        layer.iconAllowOverlap = .constant(true)
        layer.minZoom = stopZoomThreshold
        layer.symbolSortKey = .expression(Exp(.get) { StopFeaturesBuilder.shared.propSortOrderKey })
    }

    static func offsetAlertValue(index: Int) -> Value<[Double]> {
        .expression(Exp(.step) {
            Exp(.zoom)
            MapExp.shared.offsetAlertExp(closeZoom: false, index: Int32(index)).toMapbox()
            MapDefaults.shared.closeZoomThreshold
            MapExp.shared.offsetAlertExp(closeZoom: true, index: Int32(index)).toMapbox()
        })
    }

    static func offsetTransferValue(index: Int) -> Value<[Double]> {
        .expression(Exp(.step) {
            Exp(.zoom)
            MapExp.shared.offsetTransferExp(closeZoom: false, index: Int32(index)).toMapbox()
            MapDefaults.shared.closeZoomThreshold
            MapExp.shared.offsetTransferExp(closeZoom: true, index: Int32(index)).toMapbox()
        })
    }

    static func offsetPinValue() -> Value<[Double]> {
        .expression(Exp(.step) {
            Exp(.zoom)
            MapExp.shared.offsetPinExp(closeZoom: false).toMapbox()
            MapDefaults.shared.closeZoomThreshold
            MapExp.shared.offsetPinExp(closeZoom: true).toMapbox()
        })
    }
}

File: shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/map/StopLayerGenerator.kt
package com.mbta.tid.mbta_app.map

import com.mbta.tid.mbta_app.map.style.ArrayType
import com.mbta.tid.mbta_app.map.style.Exp
import com.mbta.tid.mbta_app.map.style.Interpolation
import com.mbta.tid.mbta_app.map.style.SymbolLayer
import com.mbta.tid.mbta_app.map.style.TextAnchor
import com.mbta.tid.mbta_app.map.style.TextJustify
import com.mbta.tid.mbta_app.map.style.downcastToColor
import com.mbta.tid.mbta_app.model.MapStopRoute
import com.mbta.tid.mbta_app.model.StopDetailsFilter
import com.mbta.tid.mbta_app.repositories.Settings
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.ExperimentalSerializationApi

public object StopLayerGenerator {
    internal val maxTransferLayers = 3
    public val stopZoomThreshold: Double = 11.0
    internal val busStopZoomThreshold = 12.0

    public val stopLayerId: String = "stop-layer"
    public val stopTouchTargetLayerId: String = "${stopLayerId}-touch-target"
    internal val stopLayerSelectedPinId = "${stopLayerId}-selected-pin"

    public val busLayerId: String = "${stopLayerId}-bus"
    internal val busAlertLayerId = "${stopLayerId}-bus-alert"

    internal fun getAlertLayerId(index: Int) = "${stopLayerId}-alert-${index}"

    internal fun getTransferLayerId(index: Int) = "${stopLayerId}-transfer-${index}"

    public sealed class State {
        public data object Default : State()

        public data class StopDetails(val selectedStopId: String) : State()

        public data class TripDetails(
            val selectedStopId: String?,
            val stopFilter: StopDetailsFilter?,
            val tripStops: List<String>?,
        ) : State()
    }

    public suspend fun createStopLayers(
        colorPalette: ColorPalette,
        state: State,
        settings: Map<Settings, Boolean>,
    ): List<SymbolLayer> {
        return withContext(Dispatchers.Default) {
            val sourceId = StopFeaturesBuilder.stopSourceId

            val stopLayer =
                createStopLayer(
                    id = stopLayerId,
                    colorPalette = colorPalette,
                    state = state,
                    settings = settings,
                )

            val stopTouchTargetLayer = SymbolLayer(id = stopTouchTargetLayerId, source = sourceId)
            stopTouchTargetLayer.iconImage = Exp.image(Exp(StopIcons.stopDummyIcon))
            stopTouchTargetLayer.iconPadding = 22.0
            includeSharedProps(stopTouchTargetLayer, forBus = false, state = state)
            if (settings[Settings.ShiftingIncludeStops] == true) {
                stopTouchTargetLayer.iconOffset = shiftFromLine(settings, scale = 1.0)
            }

            val stopSelectedPinLayer = SymbolLayer(id = stopLayerSelectedPinId, source = sourceId)
            stopSelectedPinLayer.iconImage =
                Exp.case(
                    MapExp.selectedExp(state) to Exp.image(Exp(StopIcons.stopPinIcon)),
                    Exp.image(Exp("")),
                )
            stopSelectedPinLayer.textField =
                Exp.case(
                    MapExp.selectedExp(state) to Exp.get(StopFeaturesBuilder.propNameKey),
                    Exp(""),
                )
            includedDefaultTextProps(stopSelectedPinLayer, colorPalette, settings)
            stopSelectedPinLayer.iconOffset = offsetPinValue(settings)
            stopTouchTargetLayer.filter =
                Exp.ge(
                    Exp.zoom(),
                    Exp.case(MapExp.isBusExp to Exp(busStopZoomThreshold), Exp(stopZoomThreshold)),
                )
            includeSharedProps(stopSelectedPinLayer, forBus = false, state = state)

            val transferLayers =
                (0.rangeUntil(maxTransferLayers)).map { index ->
                    val transferLayer =
                        SymbolLayer(id = getTransferLayerId(index), source = sourceId)
                    transferLayer.iconImage = (StopIcons.getTransferLayerIcon(index))
                    transferLayer.iconOffset = offsetTransferValue(index, settings)
                    includeSharedProps(transferLayer, forBus = false, state = state)

                    return@map transferLayer
                }

            val alertLayers =
                (0.rangeUntil(maxTransferLayers)).map { index ->
                    createAlertLayer(
                        id = getAlertLayerId(index),
                        index,
                        state = state,
                        settings = settings,
                    )
                }

            val busLayer =
                createStopLayer(
                    id = busLayerId,
                    forBus = true,
                    colorPalette,
                    state = state,
                    settings,
                )
            val busAlertLayer =
                createAlertLayer(
                    id = busAlertLayerId,
                    forBus = true,
                    state = state,
                    settings = settings,
                )

            listOf(stopTouchTargetLayer, busLayer, busAlertLayer, stopLayer) +
                transferLayers +
                alertLayers +
                listOf(stopSelectedPinLayer)
        }
    }

    private fun shiftFromLine(
        settings: Map<Settings, Boolean>,
        scale: Double = 1.0,
    ): Exp<List<Number>> {
        val routeIds = Exp.get(StopFeaturesBuilder.propRouteIdsKey)
        val mapRoutes = Exp.get(StopFeaturesBuilder.propMapRoutesKey)
        fun b(relativeScale: Double = 1.0) =
            Exp.array(
                ArrayType.Number,
                2,
                listOf(
                    MapExp.lineShiftEast(scale * relativeScale, routeIds, mapRoutes, settings),
                    Exp(0),
                ),
            )
        return if (settings[Settings.ShiftingScaleWithZoom] == true) {
            Exp.interpolate(
                Interpolation.Linear,
                Exp.zoom(),
                Exp(6) to b(0.5),
                Exp(MapDefaults.midZoomThreshold) to b(0.5),
                Exp(MapDefaults.closeZoomThreshold) to b(1.0),
            )
        } else {
            b(1.0)
        }
    }

    internal fun createAlertLayer(
        id: String,
        index: Int = 0,
        forBus: Boolean = false,
        state: State,
        settings: Map<Settings, Boolean>,
    ): SymbolLayer {
        val alertLayer = SymbolLayer(id = id, source = StopFeaturesBuilder.stopSourceId)
        alertLayer.iconImage = AlertIcons.getAlertLayerIcon(index, forBus = forBus)
        alertLayer.iconOffset = offsetAlertValue(index, settings)
        alertLayer.iconAllowOverlap = true
        includeSharedProps(alertLayer, forBus, state)

        return alertLayer
    }

    internal fun createStopLayer(
        id: String,
        forBus: Boolean = false,
        colorPalette: ColorPalette,
        state: State,
        settings: Map<Settings, Boolean>,
    ): SymbolLayer {
        val stopLayer = SymbolLayer(id = id, source = StopFeaturesBuilder.stopSourceId)
        stopLayer.iconImage = (StopIcons.getStopLayerIcon(forBus = forBus))
        stopLayer.textField = (MapExp.stopLabelTextExp(forBus = forBus, state = state))
        includedDefaultTextProps(stopLayer, colorPalette, settings)
        stopLayer.textAllowOverlap = false

        includeSharedProps(stopLayer, forBus, state)
        if (settings[Settings.ShiftingIncludeStops] == true) {
            stopLayer.iconOffset =
                shiftFromLine(settings, scale = 1.0) // TODO figure out why 1/24 doesn’t work
        }
        return stopLayer
    }

    internal fun includedDefaultTextProps(
        layer: SymbolLayer,
        colorPalette: ColorPalette,
        settings: Map<Settings, Boolean>,
    ) {
        layer.textColor = Exp(colorPalette.text).downcastToColor()
        layer.textFont = listOf("Inter Regular")
        layer.textHaloColor = Exp(colorPalette.fill3).downcastToColor()
        layer.textHaloWidth = 2.0
        layer.textSize = 13.0
        layer.textVariableAnchor =
            listOf(TextAnchor.RIGHT, TextAnchor.BOTTOM, TextAnchor.TOP, TextAnchor.LEFT)
        layer.textJustify = TextJustify.AUTO
        layer.textAllowOverlap = true
        layer.textOptional = true
        layer.textOffset = MapExp.labelOffsetExp(settings)
    }

    @OptIn(ExperimentalSerializationApi::class)
    internal fun includeSharedProps(layer: SymbolLayer, forBus: Boolean, state: State) {
        layer.iconSize = MapExp.selectedSizeExp(state)

        layer.iconAllowOverlap = true
        layer.symbolSortKey = Exp.get(StopFeaturesBuilder.propSortOrderKey)
        layer.filter =
            if (state is State.TripDetails)
                Exp.all(
                    selectedOrBeyondThresholdExp(state, forBus),
                    // we hide bus stops off this trip/route at zoom levels < 15, so we show
                    // non-bus stops, trip/route stops, or zoom > 15
                    Exp.any(
                        MapExp.listNotEq(
                            Exp.get(StopFeaturesBuilder.propMapRoutesKey),
                            listOf(Exp(MapStopRoute.BUS.name)),
                        ),
                        if (!state.tripStops.isNullOrEmpty())
                            Exp.`in`(
                                Exp.get(StopFeaturesBuilder.propIdKey),
                                Exp.array(contents = state.tripStops.map(::Exp)),
                            )
                        else if (state.stopFilter != null)
                            Exp.`in`(
                                Exp("${state.stopFilter.routeId}/${state.stopFilter.directionId}"),
                                Exp.get(StopFeaturesBuilder.propAllRouteDirectionsKey),
                            )
                        else Exp(false),
                        Exp.ge(Exp.zoom(), Exp(MapDefaults.closeZoomThreshold)),
                    ),
                )
            else selectedOrBeyondThresholdExp(state, forBus)
    }

    private fun selectedOrBeyondThresholdExp(state: State, forBus: Boolean) =
        Exp.any(
            MapExp.selectedExp(state),
            Exp.ge(Exp.zoom(), if (forBus) Exp(busStopZoomThreshold) else Exp(stopZoomThreshold)),
        )

    internal fun offsetAlertValue(index: Int, settings: Map<Settings, Boolean>): Exp<List<Number>> {
        return Exp.step(
            Exp.zoom(),
            MapExp.offsetAlertExp(closeZoom = false, index, settings),
            Exp(MapDefaults.closeZoomThreshold) to
                MapExp.offsetAlertExp(closeZoom = true, index, settings),
        )
    }

    internal fun offsetTransferValue(
        index: Int,
        settings: Map<Settings, Boolean>,
    ): Exp<List<Number>> {
        return Exp.step(
            Exp.zoom(),
            MapExp.offsetTransferExp(closeZoom = false, index, settings),
            Exp(MapDefaults.closeZoomThreshold) to
                MapExp.offsetTransferExp(closeZoom = true, index, settings),
        )
    }

    internal fun offsetPinValue(settings: Map<Settings, Boolean>): Exp<List<Number>> {
        return Exp.step(
            Exp.zoom(),
            MapExp.offsetPinExp(closeZoom = false, settings = settings),
            Exp(MapDefaults.closeZoomThreshold) to
                MapExp.offsetPinExp(closeZoom = true, settings = settings),
        )
    }
}
